### PR TITLE
Allow empty named data source. Fixes #1392

### LINKED ISSUE #1392: Add PlaceholderData interface
(For binding data at runtime) 

To allow sharing across multiple units, maybe allow `name` property?

### PR DESCRIPTION
**Pull request by domoritz**
Does not fully work yet because vega fails to compute layout when there is no data.  I consider that beyond the scope of this pr.

### CODE REVIEW

File: src/compile/model.ts [hunk 1]
```diff
@@ -271,6 +271,9 @@ export abstract class Model {
   public abstract channelHasField(channel: Channel): boolean;
 
   public getName(text: string, delimiter: string = '_') {
+    if (this.data && text === String(SOURCE) && isNamedData(this.data)) {
```
Comment (kanitw) [inline src/compile/model.ts]: Do we really need `String(SOURCE)`?  Can't it just be `SOURCE`?

**Comment by kanitw** [comment]
> Does not fully work yet because vega fails to compute layout when there is no data. I consider that beyond the scope of this pr.

Hmm.  What happens after we bind data?

File: src/compile/model.ts (hunk 1 above)
Comment (domoritz) [inline src/compile/model.ts]: Fixed

**Comment by domoritz** [comment]
> Hmm. What happens after we bind data?
🤷

**Comment by kanitw** [comment]
Actually, does the spec looks correct for layered plot too? If so, LGTM.

**Comment by kanitw** [comment]
@domoritz  FYI, I rebase and codecov is all good right now awesome!!!!

**Comment by domoritz** [comment]
I tried

Does work:
```js
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {
    "name": "foo"
  },
  "mark": "bar",
  "encoding": {
    "x": {"field": "a", "type": "ordinal"},
    "y": {"field": "b", "type": "quantitative"}
  }
}
```

Does not work (yet)
```js
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {
    "name": "foo"
  },
  "layer": [
    {
      "mark": "bar",
      "encoding": {
        "x": {"field": "a","type": "ordinal"},
        "y": {"field": "b","type": "quantitative"}
      }
    },
    {
      "mark": "line",
      "encoding": {
        "x": {"field": "a","type": "ordinal"},
        "y": {"field": "b","type": "quantitative"},
        "color": {"value": "red"}
      }
    }
  ]
}
```

**Comment by kanitw** [comment]
I guess we can merge but leave #1392 open for composition?

**Comment by kanitw** [comment]
Or if you wanna fix it now, fine.  up to you

**Comment by domoritz** [comment]
I added a note in https://github.com/vega/vega-lite/issues/1602

**Pull request merged**